Documentation/design-docs/host-startup-hook.md

```diff
@@ -0,0 +1,201 @@
+name following an exclamation mark. The list is delimited by the
+platform-specific path separator (`;` on Windows and `:` on Unix). It
+may not contain any empty entries or a trailing path separator.
+
+Setting this environment variable will cause each of the specified
+types' `public static void Initialize()` methods to be called in
+order, synchronously, before the main assembly is loaded. The hooks
+are all called on the same managed thread (the same thread that calls
+`Main`). The environment variable will be inherited by child processes
+by default. It is up to the `StartupHook.dll`s and user code to decide
+what to do about this - `StartupHook.dll` may clear them to prevent
+this behavior globally, if desired.
+
+Specifically, hostpolicy starts up coreclr and sets up a new
+AppDomain. It then invokes a private method in
+`System.Private.CoreLib`, which will call each
+`StartupHookType.Initialize()` in turn synchronously. This gives
+`StartupHookType` a chance to set up new `AssemblyLoadContext`s, or
+register other callbacks. After all of the `Initialize()` methods
+return, control returns to hostpolicy, which then calls the main entry
+point of the app like usual.
+
+Rather than forcing all configuration to be done through a single
+predefined API, this creates a place where such configuration could be
+centralized, while still allowing user code to do its own thing if it
+so desires.
+
+The producer of `StartupHook.dll` needs to ensure that
+`StartupHook.dll` is compatible with the dependencies specified in the
+main application's deps.json, since those dependencies are put on the
+TPA list during the runtime startup, before `StartupHook.dll` is
+loaded. This means that `StartupHook.dll` needs to built against the
+same or lower version of .NET Core than the app.
+
+## Example
+
+This could be used with `AssemblyLoadContext` APIs to resolve
+dependencies not on the TPA list from a shared location, similar to
+the GAC on full framework. It could also be used to forcibly preload
+assemblies that are on the TPA list from a different location. Future
+changes to `AssemblyLoadContext` could make this easier to use by
+making the default load context or TPA list modifiable.
+
+```c#
+namespace SharedHostPolicy
+{
+    class SharedHostInitializer
+    {
+        public static void Initialize()
+        {
+            AssemblyLoadContext.Default.Resolving += SharedAssemblyResolver.LoadAssemblyFromSharedLocation;
+        }
+    }
+
+    class SharedAssemblyResolver
+    {
+        public static Assembly LoadAssemblyFromSharedLocation(AssemblyLoadContext context, AssemblyName assemblyName)
+        {
+            string sharedAssemblyPath = // find assemblyName in shared location...
+            if (sharedAssemblyPath != null)
+                return AssemblyLoadContext.Default.LoadFromAssemblyPath(sharedAssemblyPath)
+            return null;
+        }
+    }
+}
+```
+
+## Error handling details
```

Comment:
More of a technical consideration: Since this happens so early on, there's no way for the program to setup any kind of exception handling. Please make sure that the experience is meaningful and document it somewhere. Not sure what .NET Core does on top-level unhandled exception, but I know that .NET Framework will trigger Watson and then write it out to console.
I guess it should be the same as what happens for top-level unhandled exception...

Reply:
I noted that the behavior is the same as for top-level exceptions in the doc, thanks!